crates/solver/src/driver.rs

```diff
@@ -384,14 +386,19 @@ impl Driver {
         let current_block_during_liquidity_fetch =
             current_block::block_number(&self.block_stream.borrow())?;
 
-        let before_count = auction.orders.len();
+        let before = auction.orders.clone();
```

Comment:
Instead of cloning all of the orders, can we already just keep the hash-set of order UIDs?

Something like:
```
let mut order_uids = auction.orders.iter().map(|order| order.metadata.uid).collect::<HashSet>();
```

And then later:
```rust
for order in &auction.orders {
    order_uids.remove(&order.metadata.uid);
}
if !order_uids.is_empty() {
    // ...
}
```

```

Reply:
I was actually quite surprised that `Order: Hash`. It feels... weird.

For example, consider the following order:
```rust
let order0 = Order {
    metadata: OrderMetadata {
       uid: x,
       executed_amount: 0,
       // ...
    },
    data: OrderData {
       partially_fillable: true,
       // ...
    },
    // ...
}
```

And
```rust
let order1 = {
    let mut order = order0.clone();
    order.metadata.executed_amount += 1; // so it got a little executed
    order
};
```

Then, intuitively, you might expect:
```rust
let mut set = HashSet::new();
set.insert(order0);
set.insert(order1);
assert!(set.len(), 1);
```

HOWEVER, this is not the case, as it would be actually:
```rust
assert!(set.len(), 2);
```

I would even go as far as to say that we should remove `derive(Hash)` for `Order`s.

Reply:
What about letting `update_and_filter()` return the removed orders?
Then we wouldn't even have to store the `before_count`.

Reply:
`update_and_filter()` taking a `&mut Auction` makes it weird but not impossible:
```rust
    pub fn update_and_filter(&mut self, auction: &mut Auction) -> Vec<Order> {
        //....
        let mut auction_orders = vec![];
        std::mem::swap(&mut auction.orders, &mut auction_orders);

        let (mut allowed_orders, in_flight_orders) =
            auction_orders.into_iter().partition(|order| {
                u256_to_big_uint(&order.data.buy_amount) > order.metadata.executed_buy_amount
                    && u256_to_big_uint(&order.data.sell_amount)
                        > order.metadata.executed_sell_amount
            });
        std::mem::swap(&mut auction.orders, &mut allowed_orders);

        in_flight_orders
    }
```

Reply:
Nice, thanks for the suggestion. This actually revealed a bug in update_and_filter (will send a separate PR, although that may become a logic involving PR 😨